feat(biome_css_analyze): add rule `useSortedProperties`

simon-paris opened the pull request:
<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

## Summary

Related: https://github.com/biomejs/biome/discussions/3#discussioncomment-10699913, https://github.com/biomejs/biome/discussions/3167

This implements CSS declaration and rule sorting from the `stylelint-order` plugin. It sorts properties (and nested rules) in a consistent order, e.g:

```
  ! Properties can be sorted.
  
  > 1 │ .class { color: red; display: block; }
       │       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  i Safe fix: Sort these properties
  
    1    │ - .class·{·color:·red;·display:·block;·}
       1 │ + .class·{·display:·block;·color:·red;·}
```


The property order comes from the most popular preset config: `stylelint-config-recess-order`.

This is the equivalent stylelint config:
```
{
	extends: ['stylelint-config-recess-order'],
	rules: {
		order/order: [
		        "custom-properties",
		        "declarations",
		        "rules",
		        "at-rules"
	        ]
	},
}
```

The sort order is this: (ties stay in the same relative order)
 1. Node kind:
     a. Custom properties
     b. The "composes" property
     c. Properties
     e. Nested rules and at-rules
     f. Other
 2. Property name:
     a. By the order defined in PROPERTY_ORDER
 3. Vendor prefix:
     a. By the order defined in VENDOR_PREFIXES
     b. Properties with no vendor prefix

Some possible issues:

 * This is the only CSS rule that has a quickfix. I'm not sure if quickfixing CSS is stable. Seems to work though.
 * ~stylelint-config-recess-order has an open PR that changes the order a little bit. It'd be best to wait for that to merge and sync the ordering before promoting this rule.~
 * ~stylelint-config-recess-order doesn't list every property. There are 230 (rarely used) properties that have no defined order.~ Some newly released properties aren't listed. I'd rather not unilaterally decide where they go so I've left them out for now. This rule silences itself if any property exists that isn't in the list. The stylelint version doesn't raise errors for unknown properties but applying any autofix will move them to the end.
 * I've had to duplicate some logic from the noShorthandPropertyOverrides rule because this rule would apply incorrect quickfixes in that case. I've made this rule silence itself if that rule would have fired. Not sure if that's the right approach. The stylelint plugin doesn't do this and applies the fix even when its unsafe.

## Test Plan

Unit tests are included.

ematipico commented:
Thank you @simon-paris for the contribution. I will look into this rule this week. As a very early feedback, this rule should be made an assist (apologies for the lack of documentation, I will try to fix that ASAP).

Here's an example of assist: https://github.com/biomejs/biome/blob/main/crates/biome_json_analyze/src/assists/source/use_sorted_keys.rs

EvHaus commented on crates/biome_configuration/src/analyzer/linter/rules.rs:
```diff
@@ -3436,6 +3436,10 @@ pub struct Nursery {
     pub use_sorted_classes:
         Option<RuleFixConfiguration<biome_js_analyze::options::UseSortedClasses>>,
+    #[doc = "Enforce ordering of CSS properties."]
+    #[serde(skip_serializing_if = "Option::is_none")]
```
I know sorting alphabetically may have issues ([example](https://github.com/prettier/prettier/issues/2502#issuecomment-315713654)) but I'd really appreciate having an option for it. For me it's much more useful than the "recess-order" configuration. Would it be difficult to add support for it?

ematipico replied:
What's a "recess-order"?

EvHaus replied:
The grouped order that `stylelint-config-recess-order` uses by default, and the one this PR was based on.

jeddy3 replied:
I'd appreciate having an option, too.

It [looks like](https://github.com/biomejs/biome/pull/4063#issue-2544682488) `stylelint-config-recess-order` was chosen because it's the [most popular](https://level222.github.io/stylelint-order-comparison/public/) shared config. However, the Stylelint order plugin [has a dedicated rule](https://github.com/hudochenkov/stylelint-order/tree/master?tab=readme-ov-file#rules) for alphabetically sorting, and that may not have been taken into account. It's difficult to compare its popularity with those of the shared configs, as they have their own npm download stats. 

The dedicated rule is used in [2.2k public configs](https://github.com/search?q=%22order/properties-alphabetical-order%22:+true&type=code), so a lot of people prefer to sort alphabetically. Probably because there are 600+ properties, with more added all the time, so it's hard to maintain meaningful categories for all these.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+    ///
+    /// This rule checks if the properties and nested rules are in a consistent order.
+    ///
+    /// The expected ordering is roughly:
+    ///  - Custom properties
+    ///  - Layout properties (display, flex, grid)
+    ///  - Margin & padding properties
+    ///  - Typography properties (font, color)
+    ///  - Interaction properties (pointer-events, visibility)
+    ///  - Background & border properties
+    ///  - Transition & animation properties
+    ///  - Nested rules
+    ///  - Nested media queries & at-rules
```
```suggestion
    /// The expected ordering is roughly:
    ///  1. Custom properties
    ///  1. Layout properties (display, flex, grid)
    ///  1. Margin & padding properties
    ///  1. Typography properties (font, color)
    ///  1. Interaction properties (pointer-events, visibility)
    ///  1. Background & border properties
    ///  1. Transition & animation properties
    ///  1. Nested rules
    ///  1. Nested media queries & at-rules
```

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+    None
+}
+
+const SORT_INFO_KIND_CUSTOM_PROPERTY: u32 = 1;
+const SORT_INFO_KIND_COMPOSES_PROPERTY: u32 = 2;
+const SORT_INFO_KIND_DECLARATION: u32 = 3;
+const SORT_INFO_KIND_UNKNOWN_DECLARATION: u32 = 4;
+const SORT_INFO_KIND_NESTED_RULE_OR_AT_RULE: u32 = 5;
```
Rust `enum` implements sorting out of the box based on their order of appearance, when using the derives `Ord` and `PartialOrd`.

This means that if you have something like this:

```rs
#[derive(Ord, PartialOrd)]
enum Order {
	First,
	Second,
	Third
}
```

`Order::First` will have a lower order than `Order::Third`. You can also go further and represent the enum with integers, *if you need*:

```rs
#[derive(Ord, PartialOrd)]
#[repr(u32)]
enum Order {
	First,
	Second,
	Third
}
```

I don't think you need these constants

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+    }
+}
+
+fn sort_info(item: &AnyCssDeclarationOrRule) -> SortInfo {
```
nit: You could this using `From`

```rs
impl From<&AnyCssDeclarationOrRule> for SortInfo {
	fn from(node: &AnyCssDeclarationOrRule) -> SortInfo {

	}
}

```

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+                .ok()
+                .and_then(|decl| decl.property().ok());
+
+            if let Some(prop) = prop {
```
You can use the `let/else` syntax to avoid too much branching:

```rs
let Some(prop) = prop else {
	return SortInfo::unknown();
};

match prop {}
```

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+    }
+}
+
+fn contains_shorthand_after_longhand(items: &[NodeWithPosition]) -> bool {
```
Can you please add some doc comments to this function, and explain what it does?

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+}
+
+fn contains_shorthand_after_longhand(items: &[NodeWithPosition]) -> bool {
+    let mut seen_shorthand_properties = HashSet::<String>::with_capacity(items.len());
```
Using `with_capacity` isn't very efficient in this case, because inside the loop there are cases of early return e.g. `return true`. This means that you create a list where some slots will be always empty, and "wasted".

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+
+    // (iterating backwards means we can store a smaller list of seen shorthands instead of seen longhands)
+    for item in items.iter().rev() {
+        let node = &item.node;
+        if let AnyCssDeclarationOrRule::CssDeclarationWithSemicolon(decl_with_semicolon) = node {
```
```suggestion
        if let AnyCssDeclarationOrRule::CssDeclarationWithSemicolon(decl_with_semicolon) = &item.node {
```

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+
+#[derive(Clone)]
+struct NodeWithPosition {
+    position: usize,
```
What does this position represent here? Without documentation, it's hard to understand. I checked where it's used, and it seems it's used in the diagnostics. Can we clarify? 

On a side note, I don't think telling users about the concept of "position" is correct. We shouldn't have it. Instead, let's save the `TextRange` of the node, so in the diagnostic we can point the user to the exact node.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+        .collect()
+}
+
+// takes the unsorted list and returns the first item that is out of order, and the item it should be placed before
+// returns None if the list is sorted
```
```suggestion
/// Takes the unsorted list and returns the first item that is out of order, and the item it should be placed before.
/// Returns [None] if the list is sorted
```

Rust has the concept of ["doc comments"](https://doc.rust-lang.org/rust-by-example/meta/doc.html#doc-comments), which support markdown and render nicely inside IDEs. Let use them :)

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+
+// takes the unsorted list and returns the first item that is out of order, and the item it should be placed before
+// returns None if the list is sorted
+fn get_first_out_of_order_pair(
+    items: &[NodeWithPosition],
+) -> Option<(NodeWithPosition, NodeWithPosition)> {
```
Check if a list is sorted can be done using the data structures that Rust gives us.

You can take inspiration from the [`use_sorted_keys`](https://github.com/biomejs/biome/blob/6ea885f0cec47a6db48d662cbd828f4f335209ba/crates/biome_json_analyze/src/assists/source/use_sorted_keys.rs#L57) rule (it's very simple).

First, your list becomes a [`BTreeSet`/`BTreeMap`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html). This data structure orders the elements of the list using the `Ord` and `PartialOrd` trait. This means that you will implement `Ord` and `PartialOrd` to the single item.

https://github.com/biomejs/biome/blob/6ea885f0cec47a6db48d662cbd828f4f335209ba/crates/biome_json_analyze/src/assists/source/use_sorted_keys.rs#L41-L49

Then, the list can call a `is_sorted` method that checks the position of the nodes:

https://github.com/biomejs/biome/blob/6ea885f0cec47a6db48d662cbd828f4f335209ba/crates/biome_json_analyze/src/assists/source/use_sorted_keys.rs#L65-L74

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+    let mut first_out_of_order_node: Option<NodeWithPosition> = None;
+
+    // find the first node that's 'less than' the previous node
+    let mut prev_item_info = sort_info(&items[0].node);
```
Don't access items of lists using indexes, unless you are 100% sure it's safe. Rust exposes safe methods to work around this.

This means that you can simplify your code like this:

```rs
let first_item = items.get(0)?;
// notice the try operator ?, which returns early if `get()` returns `None`
let mut prev_item_info = sort_info(&first_item.node);
```

And you don't need the `is_empty` check anymore

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+        return None;
+    }
+
+    let mut first_out_of_order_node: Option<NodeWithPosition> = None;
```
```suggestion
    let mut first_out_of_order_node = None;
```

You don't need an explicit type. Rust compiler is very good at inferring types, so you usually want to specify types only when the compiler tells you to do so.

ematipico commented on crates/biome_css_analyze/src/keywords.rs:
```diff
@@ -5422,6 +5424,25 @@ pub const LONGHAND_SUB_PROPERTIES_OF_SHORTHAND_PROPERTIES: [&[&str]; 57] = [
 ];
 
+// longhand => shorthand map
+pub(crate) static LONGHAND_SUB_PROPERTIES_MAP: LazyLock<HashMap<String, String>> =
+    LazyLock::new(|| {
+        let cap: usize = LONGHAND_SUB_PROPERTIES_OF_SHORTHAND_PROPERTIES
+            .map(|list| list.len())
+            .iter()
+            .sum();
+        let mut map: HashMap<String, String> = HashMap::with_capacity(cap);
+        for (shorthand, longhand_list) in SHORTHAND_PROPERTIES
+            .iter()
+            .zip(LONGHAND_SUB_PROPERTIES_OF_SHORTHAND_PROPERTIES)
+        {
+            for longhand in longhand_list {
+                map.insert((*longhand).to_string(), (*shorthand).to_string());
+            }
+        }
+        map
+    });
```
This isn't very efficient, and we should not have it.

We are basically allocating a `String` for a string that we already have in memory. 

Considering how you use this constant, I think it's best to have a function called `is_longhand_sub_property` in the rule, which checks `SHORTHAND_PROPERTIES` and `LONGHAND_SUB_PROPERTIES_OF_SHORTHAND_PROPERTIES`

This comment is valid for `RESET_TO_INITIAL_PROPERTIES_MAP` too

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+// returns (vendor prefix, name without vendor prefix) in lowercase for a CssIdentifier
+fn css_identifier_to_prop_name(ident: &CssIdentifier) -> Option<(String, String)> {
+    let tok = ident.value_token().ok()?;
+    let raw_prop_name = tok.token_text_trimmed().text().to_string();
```
Don't allocate a string with `to_string()`

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+}
+
+// returns (vendor prefix, name without vendor prefix) in lowercase for a CssIdentifier
+fn css_identifier_to_prop_name(ident: &CssIdentifier) -> Option<(String, String)> {
```
Let's not allocate `String` if we don't use them in diagnostics/messages.

Considering how you use this information, it's best to return a `&str` or a `&'static str`

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,409 @@
+const SORT_INFO_KIND_NESTED_RULE_OR_AT_RULE: u32 = 5;
+
+// this struct can be directly sorted using the default cmp implementation
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct SortInfo {
+    kind: u32,
+    property: u32,
+    vendor_prefix: u32,
+}
```
I'm not sure if we still need data structure after the review, but if we need, you should document all fields, so developers understand what they are for.

ematipico reviewed (changes requested):
Sorry @simon-paris for the late, very late review. 

I finally managed to understand your code and look at it. There's a lot of room for improvement. I suggest to check how [`use_sorted_keys`](https://github.com/biomejs/biome/blob/6ea885f0cec47a6db48d662cbd828f4f335209ba/crates/biome_json_analyze/src/assists/source/use_sorted_keys.rs) is implemented, and emulate the same behaviour/implementation.

simon-paris commented:
Thanks for the feedback!

I've addressed most of the issues already, I'll finish it up on monday.

simon-paris commented:
Alrighty everything is addressed, ready for re-review.

Main changes:
1. Changed sorting to use a btree, structured code similar to use_sorted_keys
2. Replaced the complicated diagnostic message with a simple one
3. Removed all string allocations

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+                .as_sorted(&state.original_properties)?,
+        );
+
+        return Some(RuleAction::<CssLanguage>::new(
+            ActionCategory::QuickFix(Cow::Borrowed("")),
+            ctx.metadata().applicability(),
+            markup! { "Sort these properties" }.to_owned(),
+            mutation,
+        ));
```
```suggestion
        return Some(CssRuleAction::new(
            ActionCategory::QuickFix(Cow::Borrowed("")),
            ctx.metadata().applicability(),
            markup! { "Sort these properties" }.to_owned(),
            mutation,
        ));
```

You probably missed some information from out [contribution guide](https://github.com/biomejs/biome/blob/eedb22edf7ec27d847ca31d0c101f8fa5b4956fd/crates/biome_analyze/CONTRIBUTING.md?plain=1#L244-L258). Can you please update the code accordingly? The first argument is wrong

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+        return Some(RuleDiagnostic::new(
+            rule_category!(),
+            state.block.range(),
+            markup! {
+                "Properties can be sorted."
+            },
```
Can you please add a second message? Rule diagnostics must follow our [rule pillars](https://biomejs.dev/linter/#rule-pillars). In this case, we need to explain why the sorting is good. We can probably say something like "Having a consistent order or CSS properties can improve the readability"

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+            .items()
+            .into_iter()
+            .collect::<Vec<AnyCssDeclarationOrRule>>();
+        let is_unsafe_to_sort = contains_shorthand_after_longhand(&original_properties)
+            || contains_unknown_property(&original_properties);
```
If `is_unsafe_to_sort` is `true` we can bail right away. This improve the performance because we don't need to create `RecessOrderProperties` early, which is the type that has the bulk of the logic

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug)]
+/// Define sort order using lexographical sorting of this struct
+pub struct SortInfo {
+    // First, nodes are sorted by the kind of node (e.g. declarations before at-rules)
```
```suggestion
    /// First, nodes are sorted by the kind of node (e.g. declarations before at-rules)
```
I'm not really sure this comment belongs to `kind`. I think it belongs to `enum NodeKindOrder`, and it should be more detailed. 

Plus, add a normal comment (not doc comment `///`) with a note that the order is important in this enum. Also, it would be helpful to document each variant with example of properties

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+pub struct SortInfo {
+    // First, nodes are sorted by the kind of node (e.g. declarations before at-rules)
+    kind: NodeKindOrder,
+    // Property nodes are sorted by a predefined desired order
+    property: u32,
```
```suggestion
    /// Property nodes are sorted by a predefined desired order
    property: u32,
```

This comment should actually explain what's used for. In this case, I'm not even sure what "desired order" means in this context

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+    property: u32,
+    // Vendor prefixed properties should be before the non-prefixed version of the same property
+    vendor_prefix: u32,
+    original_position: usize,
```
Documentation is missing. We should explain what's used for. Also, as mentioned before, when we talk about "position" we usually use `TextRange`. However, it seems that here we have a semantic that isn't documented, so make sure to explain what this is and how it's used. 

From the code, this seems to be an index, right? If I were you I would try to rely to `TextRange`, because you can't go wrong with that, instead indexes aren't unique

simon-paris replied:
The original_position has two purposes -
 * Make the sort stable, I could use TextRange to achieve that.
 * Lookup index into original array so `as_sorted` can copy the nodes into the new order, I can't use TextRange for that.

I've left a comment explaining it. Let me know if this isn't the right approach.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+    }
+}
+
+fn css_identifier_to_prop_text(ident: &CssIdentifier) -> Option<TokenText> {
+    let tok = ident.value_token().ok()?;
+    Some(tok.token_text_trimmed())
+}
```
We don't need this function, since it's used only in one place.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+        .as_css_generic_property()?
+        .name()
+        .ok()
+        .and_then(|prop_name| -> Option<TokenText> {
```
```suggestion
        .and_then(|prop_name| {
```

No need to add a type, Rust compiler is very good at inferring types. Usually, you want to add types only when the compiler tells you to.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+}
+
+/// Returns a declaration's vendor prefix in lowercase
+fn prop_text_to_prefix(tok_text: &TokenText) -> Option<&'static str> {
+    let prop_lowercase = tok_text.text().to_lowercase_cow();
+    let prefix = vender_prefix(&prop_lowercase);
+    if prefix.is_empty() {
+        return None;
+    }
+    Some(prefix)
+}
```
We have a crate called `biome_string_case` to handle casing. Please use this crate, as it handles everything for you.

simon-paris replied:
I'm using `to_lowercase_cow` from `biome_string_case`, do you mean for it to use `Case::convert` instead?

Using `Case::convert` would allocate a lot of `String`. The current implementation only allocates when the string is not already lowercase.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+
+/// Returns a declaration's property name without vendor prefix in lowercase.
+fn prop_text_to_unprefixed(tok_text: &TokenText) -> Cow<'_, str> {
+    let prop_lowercase = tok_text.text().to_lowercase_cow();
```
Same here. Use `biome_string_case`

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+    // Property nodes are sorted by a predefined desired order
+    property: u32,
+    // Vendor prefixed properties should be before the non-prefixed version of the same property
+    vendor_prefix: u32,
```
Here's an important information that I managed to understand from the code, but it isn't explained here: `vendor_prefix` and `property` are the indexes extracted from two arrays defined somewhere else. I believe we should at least document this **important** peace of info, because if we change the arrays in any way, this rule changes too.

Also, let's document where the two arrays. e.g. `property` index comes from `PROPERTY_ORDER_MAP`. Also, I would change the names of the properties to have `_index` in their names.

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+                    .ok()
+                    .and_then(|decl| decl.property().ok());
+
+                match prop {
+                    Some(AnyCssProperty::CssComposesProperty(_)) => {
+                        SortInfo::from_kind(NodeKindOrder::ComposesProperty)
+                    }
+                    Some(AnyCssProperty::CssGenericProperty(prop)) => match prop.name().ok() {
+                        Some(AnyCssDeclarationName::CssDashedIdentifier(_)) => {
+                            SortInfo::from_kind(NodeKindOrder::CustomProperty)
+                        }
+                        Some(AnyCssDeclarationName::CssIdentifier(ident)) => {
```
nit: To reduce branching and avoid writing `Some()` every time, you can use `let/else` e.g. 

```rs
let Some(prop) = prop else {
	return SortInfo::unknown()
};

match prop {
	AnyCssProperty::CssComposesProperty(_) => {},
	_ => {} // add the rest
}
```

ematipico commented on crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs:
```diff
@@ -0,0 +1,387 @@
+                                let vendor_prefix_idx = match prefix {
+                                    Some(prefix) => VENDOR_PREFIXES
+                                        .iter()
+                                        .position(|vp| vp == &prefix)
+                                        .map_or(u32::MAX, |pos| pos as u32),
```
Do we need to use `u32`? Can't we use `usize`?

ematipico commented on crates/biome_css_analyze/src/order.rs:
```diff
@@ -0,0 +1,486 @@
+    font-smooth, text-outline, text-overflow-ellipsis, text-overflow-mode, nav-index, color-profile
+*/
+pub(crate) const PROPERTY_ORDER: [&str; 370] = [
+    // special
```
It would be nice to explain why it's special

ematipico reviewed (commented):
Second review. Main points:
- Technical documentation is missing, we should strive to set on stone as much doc as possible
- User documentation is not enough, we should provide more exhaustive examples
- The diagnostic should be slightly improved to match our rule pillars
- The code, generally, can be simplified

simon-paris commented:
@ematipico Thanks! I've addressed everything except the biome_string_case issue. 

Main changes:
 * Added better documentation and comments
 * Added note to diagnostic, fixed ActionCategory parameter
 * Use `let else` to reduce code nesting in several places

Thank you for your patience 🙏

simon-paris commented on crates/biome_css_analyze/src/order.rs:
```diff
@@ -0,0 +1,486 @@
+    //             )
+    //         }
+    //     }
+    // }
```
I've fixed this test instead

simon-paris commented on crates/biome_css_analyze/src/order.rs:
```diff
@@ -0,0 +1,486 @@
+    //                 with_end
+    //             );
+    //         }
+    //     }
```
I've fixed this test instead too

ematipico commented:
Yeah `let/else` is very powerful and useful. It's a new syntax that came out not long ago, so we recently started using it.

simon-paris commented:
I've made some further changes:
 * Use a handwritten Ord implementation instead of using the default lexicographical implementation.
 * Update the ordering to match stylelint-config-recess-order [v6](https://github.com/stormwarning/stylelint-config-recess-order/blob/main/docs/decisions/003-order-guidelines.md#decision). This groups properties by their spec module.

ematipico reviewed (approved):
Thank you @simon-paris for your great contribution! I will make sure to move the rule in the right place after the merge :)
